examples/usercmodule/subpackage/modexamplepackage.c

```diff
@@ -0,0 +1,75 @@
+// Include MicroPython API.
+#include "py/runtime.h"
+
+// Define example_package.foo.bar.f()
+STATIC mp_obj_t example_package_foo_bar_f(void) {
+    mp_printf(&mp_plat_print, "example_package.foo.bar.f\n");
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_0(example_package_foo_bar_f_obj, example_package_foo_bar_f);
+
+// Define all properties of the second-level sub-package (example_package.foo.bar).
+STATIC const mp_rom_map_elem_t example_package_foo_bar_globals_table[] = {
+    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_example_package_dot_foo_dot_bar) },
+    { MP_ROM_QSTR(MP_QSTR_f), MP_ROM_PTR(&example_package_foo_bar_f_obj) },
+};
+STATIC MP_DEFINE_CONST_DICT(example_package_foo_bar_globals, example_package_foo_bar_globals_table);
+
+// Define example_package.foo.bar module object.
+const mp_obj_module_t example_package_foo_bar_user_cmodule = {
+    .base = { &mp_type_module },
+    .globals = (mp_obj_dict_t *)&example_package_foo_bar_globals,
+};
+
+// Define example_package.foo.f()
+STATIC mp_obj_t example_package_foo_f(void) {
+    mp_printf(&mp_plat_print, "example_package.foo.f\n");
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_0(example_package_foo_f_obj, example_package_foo_f);
+
+// Define all properties of the first-level sub-package (example_package.foo).
```

Comment:
More brainstorming ideas: instead of adding submodules to the globals dict of a module, could we introduce a new `mp_type_package` type that is the same a `mp_type_module` but contains an extra pointer to a table of submodles? Or instead of creating a new type, use a dunder like `__package_dict__` in the globals table so that the submodules can't be accessed directly?

Reply:
>  introduce a new mp_type_package type that is the same a mp_type_module

We were just talking about doing exactly this in the context of #11466... by introducing a module type "module with delegation" we can solve the weirdness around the delegation entry. This does lead to a slightly weird `type(sys) != type(os)` for example. But there are more direct ways to solve that particular issue (e.g. wrap the pointer in a dummy object).

@dpgeorge For module delegation did we consider having a ROM table of (pointer-to-module-obj, delegation function) that could be searched? (Could use a variant of `MP_REGISTER_MODULE`, i.e. `MP_REGISTER_MODULE_WITH_DELEGATION`). We could do something similar for subpackages I guess.

> use a dunder like __package_dict__ in the globals table so that the submodules can't be accessed directly?

I guess what you're saying is that module_attr would know about package_dict. That could work! Has a RAM cost of course, so that should be considered.

I guess the question is -- is this a theoretical or a practical issue for pybricks? i.e. is it really important to delay initialisation of submodules until they're explicitly imported? Can you solve this in other ways, e.g. do on-demand initialisation when methods of the submodule are used? That's extra code size of course to call `check_init()` at the start of every function in the submodule (and make_new on each type).

Reply:
> is this a theoretical or a practical issue for pybricks?

No, not particularly. I'm just throwing these out as brainstorming ideas since there isn't an obvious ideal solution. There are certainly plenty of ways to get what we need regardless of what lands upstream here.

Reply:
Let's leave this discussion/issue regarding more sophisticated `__init__` of submodules for a later time.  It's a change/improvement that can be made independently.

> For module delegation did we consider having a ROM table of (pointer-to-module-obj, delegation function) that could be searched?

No I didn't consider that.  If there were a lot of modules in that table it would be slow to search (but might be a good solution if only a few modules needed delegation).